rs/execution_environment/src/canister_manager.rs

```diff
@@ -1780,6 +1796,27 @@ impl CanisterManager {
             CanisterState::new(system_state, new_execution_state, scheduler_state);
         let new_memory_usage = new_canister.memory_usage();
 
+        // If the snapshot was uploaded, make sure the snapshot's memory hook status matches the actual status.
+        // Otherwise, the snapshot is invalid.
+        if snapshot.source() == SnapshotSource::MetadataUpload {
+            let hook_condition = new_canister.is_low_wasm_memory_hook_condition_satisfied();
+            let snapshot_hook_status = snapshot.execution_snapshot().on_low_wasm_memory_hook_status;
+            match (hook_condition, snapshot_hook_status) {
+                (true, Some(OnLowWasmMemoryHookStatus::ConditionNotSatisfied))
+                | (false, Some(OnLowWasmMemoryHookStatus::Ready))
+                | (false, Some(OnLowWasmMemoryHookStatus::Executed)) => {
```

Comment:
One more suggestion: we could move this logic into `impl OnLowWasmMemoryHookStatus` so that potential future refactorings of `OnLowWasmMemoryHookStatus` take this logic into account.